### PR TITLE
Add get dot and dot type for dvv style causality

### PR DESCRIPTION
**Pull request by russelldb**
Add's `get_entry/2` to the API, this returns the `{Actor, {Counter, TS}}` entry for the given actor and clock. It is used by riak_kv for finer grained causal tracking.

### CODE REVIEW

File: src/vclock.erl [hunk 1]
```diff
@@ -131,6 +135,21 @@ get_timestamp(Node, VClock) ->
 	false           -> undefined
     end.
 
+% @doc Get the {counter, timestamp} entry value in a VClock set from Node.
+-spec get_entry(Node :: vclock_node(), VClock :: vclock()) -> {ok, dot()} | undefined.
+get_entry(Node, VClock) ->
+    case lists:keyfind(Node, 1, VClock) of
+	false -> undefined;
```
Comment (Vagabond) [inline src/vclock.erl]: Indenting looks messed up here.

File: src/vclock.erl [hunk 2]
```diff
@@ -327,4 +346,23 @@ merge_same_id_test() ->
     ?assertEqual([{<<"1">>, {1, 3}},{<<"2">>,{1,4}},{<<"3">>,{1,5}}],
                  vclock:merge([VC1, VC2])).
 
+get_entry_test() ->
+    VC = vclock:fresh(),
+    VC1 = increment(c, increment(b, increment(a, VC))),
```
Comment (Vagabond) [inline src/vclock.erl]: Just for completeness, it might be nice to increment one of the actors more than once.

**Comment by Vagabond** [comment]
I'm pretty much OK with this PR, +1 once you've addressed the above comments.

File: src/vclock.erl (hunk 1 above)
Comment (russelldb) [inline src/vclock.erl]: Weird, not to me. /me shrugs
Oh now I see it…huh.

**Comment by russelldb** [comment]
Addressed in commit above.

File: src/vclock.erl [hunk 3]
```diff
@@ -131,6 +135,21 @@ get_timestamp(Node, VClock) ->
 	false           -> undefined
     end.
 
+% @doc Get the {counter, timestamp} entry value in a VClock set from Node.
```
Comment (seancribbs) [inline src/vclock.erl]: `lists:keyfind/3` is going to return the entire tuple -- i.e. `{Node, {Counter, Timestamp}}` -- so your spec is correct but the doc is not.

File: src/vclock.erl [hunk 4]
```diff
@@ -131,6 +135,21 @@ get_timestamp(Node, VClock) ->
...
+-spec get_entry(Node :: vclock_node(), VClock :: vclock()) -> {ok, dot()} | undefined.
+get_entry(Node, VClock) ->
+    case lists:keyfind(Node, 1, VClock) of
+        false -> undefined;
+        Entry -> {ok, Entry}
+    end.
+
+%% @doc is the given argument a valid dot, or entry?
+-spec valid_entry(dot()) -> boolean().
+valid_entry({_, {_, _}}) ->
```
Comment (seancribbs) [inline src/vclock.erl]: Don't the counter and timestamp have specific types? Could a guard be employed here for safety?

**Comment by seancribbs** [comment]
I mentioned it in the dialyzer errors in basho/riak_kv#746, but this should be able to construct a vclock from a dot, e.g.

``` erlang
-spec from_dot(dot()) -> vclock().
from_dot(Dot) -> [Dot]. %% should probably have a guard too
```

**Comment by russelldb** [comment]
I went with a new function 'desends_dot/2/'. Please to be reviewing.

**Comment by seancribbs** [comment]
:+1:

**Comment by seancribbs** [comment]
If you have time, adding the new calls to vclock_eqc might be nice.

**Comment by russelldb** [comment]
Is this good to merge @seancribbs ? I can't merge https://github.com/basho/riak_kv/pull/746 without it.

**Comment by seancribbs** [comment]
Yes, looks great. :+1: to merge

**Pull request merged**